### ISSUE TITLE
Help - getting PulseAudio to work

### ISSUE THREAD

**Issue by jkwim** [owntone/owntone-server#1269]
I tried 28.1 on a Ubuntu 20.04LTS VirtualBox VM.
With default ALSA configuration everything worked as expected.

I had PulseAudio also installed and following PulseAudio related log entry is seen. 

```
[2021-06-09 11:43:55] [  LOG]     main: - Pulseaudio

[2021-06-09 11:43:55] [ WARN]   laudio: Failed to probe ALSA card=0 - No such file or directory
[2021-06-09 11:43:55] [ INFO]   laudio: Adding ALSA device 'default' with name 'Computer'
[2021-06-09 11:43:55] [DEBUG]   laudio: Pulseaudio context state changed to ready
[2021-06-09 11:43:55] [DEBUG]   laudio: Callback for Pulseaudio sink 'alsa_output.pci-0000_00_05.0.analog-stereo' (id 0)
[2021-06-09 11:43:55] [  LOG]   laudio: Adding Pulseaudio sink 'Built-in Audio Analog Stereo' (alsa_output.pci-0000_00_05.0.analog-stereo) with name 'Computer'

[2021-06-09 11:44:32] [DEBUG]   player: Removing callback to device_streaming_cb, id 0
[2021-06-09 11:44:32] [DEBUG]   player: Registered callback to device_flush_cb with id 0 (device 0x7fc428006030, Computer)
[2021-06-09 11:44:32] [DEBUG]   player: Number of active callbacks: 1
[2021-06-09 11:44:32] [DEBUG]   laudio: Pulseaudio flush
[2021-06-09 11:44:32] [DEBUG]   laudio: Callback for Pulseaudio subscribe (id 0, event 16)
[2021-06-09 11:44:32] [DEBUG]   laudio: Callback for Pulseaudio sink 'alsa_output.pci-0000_00_05.0.analog-stereo' (id 0)
[2021-06-09 11:44:32] [DEBUG]   player: Callback request received, id is 0
[2021-06-09 11:44:32] [DEBUG]   player: Making deferred callback to device_flush_cb, id was 0
[2021-06-09 11:44:32] [DEBUG]   player: Callback from Pulseaudio device Computer to device_flush_cb (status 2)
[2021-06-09 11:44:32] [DEBUG]   player: Registered callback to device_streaming_cb with id 0 (device 0x7fc428006030, Computer)
[2021-06-09 11:44:32] [DEBUG]   player: Number of active callbacks: 1
[2021-06-09 11:44:32] [DEBUG]     main: Command has 0 pending events

[2021-06-09 11:44:32] [DEBUG]   player: Player status: paused
[2021-06-09 11:44:37] [DEBUG]   laudio: Callback for Pulseaudio subscribe (id 0, event 16)
[2021-06-09 11:44:37] [DEBUG]   laudio: Callback for Pulseaudio sink 'alsa_output.pci-0000_00_05.0.analog-stereo' (id 0)
[2021-06-09 11:44:42] [DEBUG]   player: Removing callback to device_streaming_cb, id 0
[2021-06-09 11:44:42] [DEBUG]   player: Registered callback to device_streaming_cb with id 0 (device 0x7fc428006030, Computer)
[2021-06-09 11:44:42] [DEBUG]   player: Number of active callbacks: 1
[2021-06-09 11:44:42] [DEBUG]   laudio: Pulseaudio stopping 'alsa_output.pci-0000_00_05.0.analog-stereo'
[2021-06-09 11:44:42] [DEBUG]   player: Callback request received, id is 0
[2021-06-09 11:44:42] [DEBUG]   player: Unsubscription request for quality 44100/16/2 (now 0 subscribers)
[2021-06-09 11:44:42] [DEBUG]   player: Making deferred callback to device_streaming_cb, id was 0
[2021-06-09 11:44:42] [ INFO]   player: The Pulseaudio device 'Computer' stopped
[2021-06-09 11:44:42] [DEBUG]   player: Status update - status: 3, events: 8, caller: device_streaming_cb

```

All these messages led me to believe that everything will work properly if I change the configuration to use 'pulseaudio' instead of 'alsa'.

```

        # Type of the output (alsa, pulseaudio, dummy or disabled)
#       type = "alsa"
type = "pulseaudio"

```

However then no audio output was observed.

Looked at PulseAudio Volume Control GUI:
I cannot see any application name in the Playback tab suggesting that PulseAudio really does not see "OwnTone" app.

If I stop OwnTone and run RythmBox (comes default with UbuntuDektop) immediately the app appears in Playback tab.

It will be the same with players like VLS (based  on  past experience). The apps which uses PulseAudio will appear in Playback tab.

So 
1. No sound output
2. No display on PulseAudio Volume Control applet
suggests that something is wrong in my configuration.

Please help me.

The only change that I have done w.r.t. OwnTone is changing a single line in the configuration.

I have all other requirements implemented:
1. PulseAudio starts in system mode using a systemctl script

2. /etc/group has correct entries
```
audio:x:29:pulse
pulse:x:128:
pulse-access:x:129:owntone
```
(note as per PulseAudio documentation, in Ubuntu the user 'owntone' need not be in audio group. Nevertheless I tried with and without both but same results)

3. Starting owntone as a non-root user
`owntone@jubuntu:~$ /usr/sbin/owntone -P /var/run/owntone/owntone.pid -d 5`
(note that due to permission setting difficulties the default location of PID file needed to be changed. Wish that there was an option in the configuration file for this)

Reverting the only change on OwnTone configuration brings back the system to working setup using ALSA sound.

**Comment by ejurgensen** [comment]
I tried just now with Ubuntu 20.04 and I observe the same. Well more or less, I also get a bunch of overrun warnings in the log. You are right that OwnTone should show in the playback control.

I don't have any good suggestions on what you can do, except maybe try network mode instead of system mode (haven't tried myself if it works better).

The reason OwnTone doesn't just work like Rhythmbox is that it is intended to run as a system service, not via a desktop user. Pulseaudio is very much geared towards the latter, as I recall.

I guess I have to dive back into Pulseaudio land to find out how to make this work in a better way.

**Comment by LordMyschkin** [comment]
Did you set 
server = "localhost"
in owntone.conf ?

audio {
	# Name - used in the speaker list in Remote
	nickname = "Computer"

	# Type of the output (alsa, pulseaudio, dummy or disabled)
	type = "pulseaudio"

	# For pulseaudio output, an optional server hostname or IP can be
	# specified (e.g. "localhost"). If not set, connection is made via local
	# socket.
	server = "localhost"

**Comment by ejurgensen** [comment]
I think this issue must be due to some change in Ubuntu 20, because I just tried with a virgin Ubuntu 18 VM, and it worked straight out of the box. All that was required was changing type = "pulseaudio", no changes to Pulseaudio config or anything else. That was, however, with OwnTone running as root, so I did cheat a little bit there.

**Comment by jkwim** [comment]
This confuses me - so the audio type was set to alsa? I thought you already had that working? And I thought you wanted to go for Pulseaudio?

Yes, I wanted to see the audio at PulseAudio level.

Probably what is happening is that you handover the audio to ALSA. PulseAudio anyway sits on ALSA. So the Audio is appearing in PulseAudio as well. 

Can you confirm how you handover the audio based on the the config file setting?

I am assuming that if PulseAudio sees the audio through PulseAudio Volume Control, any other app accessing PulseAudio as an input source will get the Audio as well. That is what I want to test in the long run.

**Comment by jkwim** [comment]
The next question somebody would ask then is why send audio directly to PulseAudio. I suppose PulseAudio can do audio routing gracefully without contention at the hardware level. That is if two apps want to send audio to the same HW Device PulseAudio can handle that without blocking HW device for one app or the other. This is my rough understanding of PulseAudio advantage.

**Comment by ejurgensen** [comment]
When you select alsa in the config, audio is written via the [alsa pcm api](https://www.alsa-project.org/alsa-doc/alsa-lib/pcm.html), and when you select pulseaudio it is [Pulseaudio's api](https://freedesktop.org/software/pulseaudio/doxygen/) that is used. As I recall, the alsa pcm api doesn't necessarily mean that audio is written directly to the device. On Ubuntu, it would be more of a virtual device that is in fact exposed by the Pulseaudio server, which takes care of mixing etc if multiple apps are writing via the alsa api, and then Pulseaudio uses alsa itself to write the mixed audio to the real device. But it's been a while since I looked at this, so don't trust me on this ...

**Comment by jkwim** [comment]
Why I am trying to play with this is that there are two speakers I can see now on web interface.

I can control them independently to adjust the volume or enable/disable.

1. Built-in Audio Analog Stereo
2. Computer 

1 looks like coming from PulseAudio, 2 obviously is owntone config.

Previously I only saw Computer (when running as root).

**Comment by jkwim** [comment]
![2021-06-11 03_45_14-Clipboard](https://user-images.githubusercontent.com/84637540/121605484-186be380-ca6a-11eb-9030-dbc70d9370d7.png)

**Comment by p3ck** [comment]
And then there is pipewire which offers a pulseaudio API as well as it's
own..

**Comment by jkwim** [comment]
![2021-06-11 03_59_01-Clipboard](https://user-images.githubusercontent.com/84637540/121605533-320d2b00-ca6a-11eb-99e7-2b978c36bb44.png)

**Comment by jkwim** [comment]
![2021-06-11 03_59_34-Clipboard](https://user-images.githubusercontent.com/84637540/121605589-4d783600-ca6a-11eb-8264-feda11e675ae.png)

**Comment by jkwim** [comment]
![2021-06-11 04_00_36-Clipboard](https://user-images.githubusercontent.com/84637540/121605659-6bde3180-ca6a-11eb-9154-2b5916d55578.png)

**Comment by jkwim** [comment]
All these observations while config file remaining set to ALSA and running the app as user 'owntone'.

No touching the default audio setups either ALSA or PulseAudio in Ubuntu.

**Comment by jkwim** [comment]
BTW the above dual speaker control setup could have been due to some transient state of things. After rebooting I could not reproduce that anymore.
----------------------------------------------------------------------------------------------

I was playing more with this over the last two days and here are my observations  :

I had compiled the SW with `--with-pulseaudio` since the beginning.

By relocating the log file location (via config file) and relocating the PID file (via command line option) the access to root only locations can be eliminated.
So starting and stopping can be done with a simple shell script running in user's home directory.

I tried starting during booting via systemctl. Form process starting wise it works. 
However access to PulseAudio server is an issue when you do it this way (if PulseAudio is running as a user service).

Tried systemctl process as a user also but same result with PulseAudio.

Tested the setup by logging out of Desktop and accessing via SSH from an external network and starting/stopping OwnTone.
Worked as expected!

So I suppose the way forward would be to login remotely and login and start it.

One more option remains to be tested that is to write a script which checks for presence of process and if not present start it. Then setup a cron job to invoke the script periodically. 

So now we have two ways to run it. Either modify the PulseAudio service to be on System Mode or not touch PulseAudio setup but take some extra steps to run it under an own user.

I will do a write up of the extra steps that we need to do post install to get things working as a user.

Note that this setup is with Audio being fed in to ALSA and controlled using PulseAudio.

**Comment by jkwim** [comment]
This is the flow of audio related messages in the log:

```
[2021-06-12 21:04:21] [ INFO]   laudio: Available ALSA playback mixer(s) on hw:0 CARD=I82801AAICH (Intel 82801AA-ICH): 'Master' 'Master Mono' 'PCM' 'Surround' 'Center' 'LFE' 'Line' 'CD' 'Mic' 'Video' 'Phone' 'Beep' 'Aux'
[2021-06-12 21:04:21] [ INFO]   laudio: Adding ALSA device 'default' with name 'Computer'
[2021-06-12 21:04:21] [DEBUG]   laudio: Pulseaudio context state changed to ready
[2021-06-12 21:04:21] [DEBUG]   laudio: Callback for Pulseaudio sink 'alsa_output.pci-0000_00_05.0.analog-stereo' (id 0)
[2021-06-12 21:04:21] [  LOG]   laudio: Adding Pulseaudio sink 'Built-in Audio Analog Stereo' (alsa_output.pci-0000_00_05.0.analog-stereo) with name 'Computer'
[2021-06-12 21:05:31] [DEBUG]   laudio: Pulseaudio starting 'Computer'
[2021-06-12 21:05:31] [DEBUG]   laudio: Opening Pulseaudio stream to 'alsa_output.pci-0000_00_05.0.analog-stereo'
[2021-06-12 21:05:31] [DEBUG]   laudio: Callback for Pulseaudio subscribe (id 0, event 16)
[2021-06-12 21:05:31] [DEBUG]   laudio: Callback for Pulseaudio sink 'alsa_output.pci-0000_00_05.0.analog-stereo' (id 0)
[2021-06-12 21:05:43] [DEBUG]   laudio: Setting Pulseaudio volume for stream 0 to 36044 (55)
[2021-06-12 21:05:45] [DEBUG]   laudio: Setting Pulseaudio volume for stream 0 to 33423 (51)
[2021-06-12 21:05:51] [DEBUG]   laudio: Setting Pulseaudio volume for stream 0 to 28180 (43)
```

**Comment by jkwim** [comment]
> Really great that you are diving into this, that is surely needed.

I am ready to do a write up on getting things working in user mode in Ubuntu 20.04 LTS.

I now have successfully got things working in user mode on a Rasbberry Pi 4 with support for Bluetooth Headset/Speaker with PulseAudio.

Repeated testing shows that everything work as expected.

Currently a designated user needs to login in to a headless machine and start OwnTone manually and exit the shell. The server will run continuously until the process is killed.

I hope to setup a cron job for the user to periodically check whether the server is running and if not start it. I hope that it will work.

Let me know where to post a write up? here or Wiki?

**Comment by ejurgensen** [comment]
Good work, but do you still have type = "alsa" in the config? If it is full Pulseaudio mode, then I would suggest that you make a pull request with an update of README_PULSE.md. I think having the info there would make it easiest for other users to find.

> Currently a designated user needs to login in to a headless machine and start OwnTone manually and exit the shell.

This doesn't sound ideal, but I guess there is no other alternative when it is user mode?

Did you get around to trying network mode, like suggested above?

**Comment by jkwim** [comment]
> Did you get around to trying network mode, like suggested above?

The critical requirement here is to be able to start PulseAudio as a definitive known user.

And after writing above post, I was playing with it.

Managed to get it working like this:

1. Create a user script
2. Setup a user cronj job to get fired immediately after a boot.
3. Start PulseAudio using the script as the user
4. Start OwnTone server as the user.

With this setup after a reboot, things get started correctly.

Though I have not tested, the systemctl method of restating OwnTone app should work as well because already PulseAudio has been started correctly with correct non-root user.

```
[2021-06-16 10:17:06] [DEBUG]   laudio: Pulseaudio context state changed to ready
[2021-06-16 10:17:06] [DEBUG]   laudio: Callback for Pulseaudio sink 'alsa_output.platform-bcm2835_audio.stereo-fallback' (id 0)
[2021-06-16 10:17:06] [  LOG]   laudio: Adding Pulseaudio sink 'Built-in Audio Stereo' (alsa_output.platform-bcm2835_audio.stereo-fallback) with name 'RaspberryPi4'

```

The only situation that this script would fail is if there is another user account and that user logs in before the automatic start kicks in but that is highly unlikely situation in majority of the user cases.

I am using "pulseaudio" option in the config file.

Since the Pi 4 has Bluetooth capability, I tested that also. PulseAudio connects the Bluetooth device automatically and OwnTone recognizes the new speaker immediately, whch is great.

```
[2021-06-16 10:22:56] [DEBUG]   laudio: Callback for Pulseaudio subscribe (id 1, event 16)
[2021-06-16 10:22:56] [DEBUG]   laudio: Callback for Pulseaudio subscribe (id 1, event 0)
[2021-06-16 10:22:56] [DEBUG]   laudio: Callback for Pulseaudio sink 'bluez_sink.0C_E0_E4_C8_3E_01.a2dp_sink' (id 1)
[2021-06-16 10:22:56] [  LOG]   laudio: Adding Pulseaudio sink 'PLT_BBTSENSE' (bluez_sink.0C_E0_E4_C8_3E_01.a2dp_sink)
[2021-06-16 10:22:56] [DEBUG]   laudio: Callback for Pulseaudio sink 'bluez_sink.0C_E0_E4_C8_3E_01.a2dp_sink' (id 1)

```

![image](https://user-images.githubusercontent.com/84637540/122160448-962a5780-ce8d-11eb-9753-0c297dfbc080.png)

**Comment by jkwim** [comment]
> I would suggest that you make a pull request with an update of README_PULSE.md

Need some help on doing this as I have not done that before.

Do I have to make a fork and add it?

**Comment by ejurgensen** [comment]
There's plenty of guides here on github, but here's a short version:

```
(make a fork with the Fork button)
$ git clone https://github.com/jkwim/owntone-server.git
$ cd owntone-server
$ git config --global user.name "FIRST_NAME LAST_NAME"
$ git config --global user.email "NAME@example.com"
$ git checkout -b pulse_doc1
(edit the README_PULSE.md)
$ git commit -a
$ git push origin pulse_doc1
(go to your fork and click the pull request button)
```
'pulse_doc1' is the name of the branch where you make the changes.

**Comment by jkwim** [comment]
I have added to the README_PULSE.md the steps to automate start up with user mode.

If somebody can try this out and confirm before I make a pull request, I would appreciate.

[https://github.com/jkwim/owntone-server/blob/pulse_doc1/README_PULSE.md#user-mode-with-auto-startup-script](https://github.com/jkwim/owntone-server/blob/pulse_doc1/README_PULSE.md#user-mode-with-auto-startup-script)

**Comment by ejurgensen** [comment]
Thanks for a very good writeup. I notice step 1-8 (except 2) aren't as such related to Pulseaudio, but have more to do with running Owntone as non-root. I think I will try to automate that so that most of those steps are done by `make install`.

With regard to the startup script: Have you tried if it is possible to use `systemctl --user` (see for instance [this](https://nts.strzibny.name/systemd-user-services/))?

Btw if you use `killall owntone` then you don't need to get the pid.

**Comment by jkwim** [comment]
> With regard to the startup script: Have you tried if it is possible to use `systemctl --user` (see for instance [this](https://nts.strzibny.name/systemd-user-services/))?

I hope to play with this more. I did try certain things but need to explore more. I do not know who the bindings work in OwnTone, when starting it might try to bind in to any PulseAudio server which might exist right? Suppose another user has logged in and the login process started another PulseAudio server then we need to ensure that OwnTone does not recognize that PulseAudio server. This is where the Connection Refused error comes.
Anyway let me do more testing.

> Btw if you use `killall owntone` then you don't need to get the pid.
Yes we may do that. During testing I found an issue where there were two PulseAudio servers getting started. The reason for this, imo, is that I logged in to a shell account while cron job was still executing sleep command before starting OwnTone. `killall` will gracefully stop all PulseAudio servers which is better.

**Comment by jkwim** [comment]
BTW, some thing additional with PulseAudio. I even tested a PulseAudio tunnel to another machine and then the remote machine's speaker appears as the playback device. Unfortunately PulseAudio cuts of Bluetooth devices then.

Sending the pcm stream is not a good idea from bandwidth point of view and if the network speed isn't good enough audio drops.

However this could be a practical use on adhoc basis.

One issue was that PulseAduio drops the tunnel if it is inactive for a long period and you need to restart PulseAudio servers on both sides to reestablish the tunnel. Perhaps there is a way to prevent idle detection.

**Comment by jkwim** [comment]
> I think I will try to automate that so that most of those steps are done by `make install`.

Any possibility to add the ability to change the log file location via conf file?

**Comment by ejurgensen** [comment]
> Any possibility to add the ability to change the log file location via conf file?

Isn't that already an option?

**Comment by jkwim** [comment]
> Isn't that already an option?

Typo. I wanted to mean the PID file location.

**Comment by spiro-c** [comment]
> I have added to the README_PULSE.md the steps to automate start up with user mode.
> 
> If somebody can try this out and confirm before I make a pull request, I would appreciate.
> 
> https://github.com/jkwim/owntone-server/blob/pulse_doc1/README_PULSE.md#user-mode-with-auto-startup-script

I did try you example and is working for me on raspbery4 running Ubuntu 21.04 arm64 and on virtual machine running Ubuntu 21.04 AMD64 but i did not create new user just add access for my user to `/var/log/owntone` `/var/cache/owntone` and temp files `/usr/lib/tmpfiles.d/owntone.conf` whit my user all so i create `systemctl --user` in `~/.config/systemd/user/owntone.service` whit the following config 
```
[Unit]
Description=DAAP/DACP (iTunes), RSP and MPD server, supports AirPlay and Remote
Documentation=man:owntone(8)
#Requires=network.target local-fs.target avahi-daemon.socket
After=network-online.target

[Service]
ExecStart=/usr/sbin/owntone -f

# Constrain the upper limit of memory/swap that can be used; this prevents
# the server from consuming all system memory (in event of bug/malformed user
# curl/SMARTPL query etc) that would hang/freeze low resource and headless (ie
# RPi) machines
#
# systemd will kill the process in such an event but would be auto-restarted as
# per 'Restart' directive below
#
# Values derived from obersvations on rpi3 under load - limits are >50% above
# seen high watermarks
#
# https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html
MemoryMax=256M
MemorySwapMax=32M

# Restart, but set a limit so we don't restart indefinitely. Unfortunately,
# systemd also applies the start limits to manual restarts, so that's why the
# burst value allows for 10 restarts.
Restart=on-failure
RestartSec=5
StartLimitBurst=10
StartLimitInterval=600

[Install]
WantedBy=default.target
```
and is starting on reboot no problem  pulseaudio server is running under my user i just install it whit `sudo apt install pulseaudi` and add my user to pulse-access

**Comment by jkwim** [comment]
> I did try you example and is working for me on raspbery4 running Ubuntu 21.04 arm64 and on virtual machine running Ubuntu 21.04 AMD64 but i did not create new user just add access for my user to `/var/log/owntone` `/var/cache/owntone` and temp files `/usr/lib/tmpfiles.d/owntone.conf` whit my user all so i create `systemctl --user` service in `~/.config/systemd/user/owntone.service` whit the following config

Great!

I also learned that any of the shell commands required can be incorporated in to the systemctl service file. In our case most of the commands require `root` permission so may not be an option.

Here is an example of a service file run as 'root'
```
[Service]
Type=notify
NotifyAccess=main
ExecStart=/usr/sbin/mosquitto -c /etc/mosquitto/mosquitto.conf
ExecReload=/bin/kill -HUP $MAINPID                                   <<
Restart=on-failure
ExecStartPre=/bin/mkdir -m 740 -p /var/log/mosquitto                 <<
ExecStartPre=/bin/chown mosquitto /var/log/mosquitto                 <<
ExecStartPre=/bin/mkdir -m 740 -p /run/mosquitto                     <<
ExecStartPre=/bin/chown mosquitto /run/mosquitto                     <<

```

**Comment by spiro-c** [comment]
Just update on the user service i notice that after i logout the owntone webui is not working and then i will login to the system and check the service is running but just started 5 sec ago so i do little bit digging around and turnout that i need to enable [linger](https://www.freedesktop.org/software/systemd/man/loginctl.html#enable-linger%20USER%E2%80%A6) with `loginctl enable-linger $USER` for my user so the service start at boot
